File: resume-analyzer/backend/services/nlp_engine.py
import re
from difflib import SequenceMatcher

from utils.skills_db import COMMON_SKILLS
from models.schemas import AIAnalysisResult

class NLPService:
    def __init__(self):
        # Precompute lowercase skills once for faster matching.
        self._skills = [skill.strip().lower() for skill in COMMON_SKILLS if skill.strip()]
        self._required_sections = [
            "summary",
            "experience",
            "education",
            "skills",
            "projects",
        ]
        self._critical_sections = ["experience", "skills", "education"]
        self._action_verbs = {
            "built", "led", "managed", "designed", "developed", "implemented", "optimized",
            "improved", "launched", "scaled", "automated", "delivered", "created", "reduced",
            "increased", "spearheaded", "architected", "achieved",
        }
        self._section_headings = {
            "experience", "education", "projects", "summary", "certifications",
            "achievements", "work experience", "profile", "objective",
        }
        self._fallback_stopwords = {
            "and", "or", "with", "using", "in", "of", "the", "to", "for", "a", "an",
            "skills", "technologies", "tools", "frameworks", "languages", "proficient", "knowledge",
        }

    def _tokenize(self, text: str) -> set[str]:
        return set(re.findall(r"[a-zA-Z0-9+#.-]+", text.lower()))

    def _extract_fallback_skills(self, text: str) -> list[str]:
        text = text or ""
        lines = [line.strip() for line in text.splitlines() if line.strip()]

        # Prefer explicit skills sections when present.
        collected = []
        capture = False
        for line in lines:
            lower = line.lower().strip(" :|-\t")

            if "skill" in lower:
                capture = True
                continue

            if capture and lower in self._section_headings:
                break

            if capture:
                collected.append(line)

        # If no skills section found, take first lines as weak fallback.
        source_text = "\n".join(collected[:20]) if collected else "\n".join(lines[:30])

        candidates = []
        raw_parts = re.split(r"[,|/•;\n]+", source_text)
        for part in raw_parts:
            chunk = re.sub(r"\s+", " ", part).strip(" -:\t")
            if not chunk:
                continue

            # Keep short skill-like phrases.
            words = [w for w in re.findall(r"[A-Za-z0-9+#.-]+", chunk)]
            if not words or len(words) > 4:
                continue

            normalized = " ".join(words).lower()
            if normalized in self._fallback_stopwords:
                continue
            if any(sw in normalized.split() for sw in self._fallback_stopwords):
                # Keep tokens like 'data structures' but drop noisy phrases.
                if len(words) == 1:
                    continue

            # Must contain at least one alphabetic character.
            if not re.search(r"[a-zA-Z]", normalized):
                continue

            candidates.append(normalized)

        # Deduplicate while preserving order.
        seen = set()
        ordered = []
        for skill in candidates:
            if skill in seen:
                continue
            seen.add(skill)
            ordered.append(skill)
            if len(ordered) >= 18:
                break

        return ordered

    def extract_skills(self, text: str) -> list[str]:
        text_lower = text.lower()
        tokens = self._tokenize(text)
        skills = set()

        for skill in self._skills:
            if " " in skill:
                if skill in text_lower:
                    skills.add(skill)
            elif skill in tokens:
                skills.add(skill)

        # If dictionary match is sparse, recover skills from explicit skills text patterns.
        if len(skills) < 3:
            for fallback_skill in self._extract_fallback_skills(text):
                skills.add(fallback_skill)

        return sorted(skills)

    def calculate_similarity_score(self, resume_text: str, job_desc: str) -> float:
        resume_text = (resume_text or "").strip().lower()
        job_desc = (job_desc or "").strip().lower()

        if not resume_text or not job_desc:
            return 0.0

        # Lightweight fuzzy text similarity as a semantic approximation.
        ratio = SequenceMatcher(None, resume_text, job_desc).ratio()
        return ratio * 100

    def _score_sections(self, resume_text: str) -> tuple[float, list[str]]:
        text = (resume_text or "").lower()
        found = []
        missing = []

        for section in self._required_sections:
            pattern = rf"\b{re.escape(section)}\b"
            if re.search(pattern, text):
                found.append(section)
            else:
                missing.append(section)

        section_score = (len(found) / len(self._required_sections)) * 20
        return section_score, missing

    def _score_impact_signals(self, resume_text: str) -> tuple[float, int, int]:
        text = (resume_text or "").lower()
        tokens = self._tokenize(text)

        metric_matches = re.findall(
            r"\b(\d+%|\$\d+[kKmM]?|\d+\+?\s?(years|yrs|months|users|clients|projects|features))\b",
            text,
        )
        metric_count = len(metric_matches)

        verb_count = len(self._action_verbs.intersection(tokens))

        metrics_component = min(metric_count, 5) / 5 * 6
        verbs_component = min(verb_count, 6) / 6 * 4
        return metrics_component + verbs_component, metric_count, verb_count

    def _score_formatting(self, resume_text: str) -> tuple[float, list[str]]:
        text = resume_text or ""
        suggestions = []
        score = 0.0

        email_present = bool(re.search(r"[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}", text))
        phone_present = bool(re.search(r"(\+?\d[\d\-\s]{8,}\d)", text))
        bullet_count = len(re.findall(r"(^\s*[-*•]|\n\s*[-*•])", text, flags=re.MULTILINE))
        length = len(text)

        if email_present:
            score += 3
        else:
            suggestions.append("Add a professional email in contact details.")

        if phone_present:
            score += 2
        else:
            suggestions.append("Include a phone number for ATS completeness.")

        if 450 <= length <= 9000:
            score += 3
        elif length < 450:
            suggestions.append("Resume content is too short; add role impact and project details.")
        else:
            suggestions.append("Resume content is too long; keep it concise and ATS-friendly.")

        if bullet_count >= 4:
            score += 2
        else:
            suggestions.append("Use bullet points to improve ATS parsing and readability.")

        return score, suggestions

    def _keyword_stuffing_penalty(self, resume_text: str, job_skills: set[str]) -> tuple[float, int]:
        text = (resume_text or "").lower()
        if not text:
            return 0.0, 0

        repeated_count = 0
        for skill in job_skills:
            occurrences = len(re.findall(rf"\b{re.escape(skill)}\b", text))
            if occurrences >= 6:
                repeated_count += 1

        # Penalize excessive repetition of JD keywords (possible keyword stuffing).
        penalty = min(repeated_count * 2.0, 8.0)
        return penalty, repeated_count

    def _strict_caps(
        self,
        base_score: float,
        job_skill_count: int,
        skill_coverage: float,
        missing_sections: list[str],
        metric_count: int,
        bullet_count: int,
    ) -> tuple[float, list[str]]:
        capped_score = base_score
        cap_reasons = []

        missing_critical = [section for section in self._critical_sections if section in missing_sections]

        # Gate 1: with JD provided, low skill coverage should hard-cap ATS score.
        if job_skill_count > 0:
            if skill_coverage < 0.25:
                capped_score = min(capped_score, 45.0)
                cap_reasons.append("Very low role-skill coverage (<25%).")
            elif skill_coverage < 0.40:
                capped_score = min(capped_score, 60.0)
                cap_reasons.append("Low role-skill coverage (<40%).")

        # Gate 2: missing critical resume structure should cap readiness.
        if len(missing_critical) >= 2:
            capped_score = min(capped_score, 55.0)
            cap_reasons.append("Missing multiple critical sections (Experience/Skills/Education).")

        # Gate 3: no quantifiable impact should cap strong ATS outcomes.
        if metric_count == 0:
            capped_score = min(capped_score, 65.0)
            cap_reasons.append("No quantifiable impact metrics detected.")

        # Gate 4: very weak formatting (few bullets) should cap score.
        if bullet_count < 3:
            capped_score = min(capped_score, 70.0)
            cap_reasons.append("Insufficient bullet structure for ATS parsing.")

        return capped_score, cap_reasons

    def analyze_resume_vs_job(self, resume_text: str, job_desc: str) -> AIAnalysisResult:
        resume_text = resume_text or ""
        job_desc = job_desc or ""

        resume_skills = set(self.extract_skills(resume_text))
        job_skills = set(self.extract_skills(job_desc))
        matched_skills = list(resume_skills.intersection(job_skills))
        missing_skills = list(job_skills - resume_skills)
        semantic_score = self.calculate_similarity_score(resume_text, job_desc)

        # Strict ATS rubric (0-100): skills 45, semantic 15, sections 20, impact 10, format 10.
        if job_skills:
            skill_coverage = len(matched_skills) / len(job_skills)
            skills_component = skill_coverage * 45
        else:
            # No JD provided: apply strict baseline by detected skill breadth.
            skill_coverage = 0.0
            skills_component = min(len(resume_skills), 18) / 18 * 45

        semantic_component = (semantic_score / 100) * 15
        sections_component, missing_sections = self._score_sections(resume_text)
        impact_component, metric_count, verb_count = self._score_impact_signals(resume_text)
        format_component, formatting_suggestions = self._score_formatting(resume_text)
        bullet_count = len(re.findall(r"(^\s*[-*•]|\n\s*[-*•])", resume_text, flags=re.MULTILINE))

        stuffing_penalty, stuffed_keywords = self._keyword_stuffing_penalty(resume_text, job_skills)
        weak_impact_penalty = 4.0 if metric_count < 2 else 0.0
        weak_action_penalty = 2.0 if verb_count < 3 else 0.0
        deductions = stuffing_penalty + weak_impact_penalty + weak_action_penalty

        final_score = skills_component + semantic_component + sections_component + impact_component + format_component - deductions
        final_score, cap_reasons = self._strict_caps(
            base_score=final_score,
            job_skill_count=len(job_skills),
            skill_coverage=skill_coverage,
            missing_sections=missing_sections,
            metric_count=metric_count,
            bullet_count=bullet_count,
        )

        final_score = min(100, max(0, final_score))
        final_score = round(final_score, 1)

        score_breakdown = {
            "skills_alignment": {"score": round(skills_component, 1), "max": 45.0},
            "semantic_relevance": {"score": round(semantic_component, 1), "max": 15.0},
            "section_coverage": {"score": round(sections_component, 1), "max": 20.0},
            "impact_evidence": {"score": round(impact_component, 1), "max": 10.0},
            "formatting_quality": {"score": round(format_component, 1), "max": 10.0},
            "strictness_deductions": {"score": round(deductions, 1), "max": 14.0},
            "total": {"score": final_score, "max": 100.0},
        }

        exp_match = "Strong" if final_score >= 85 else "Moderate" if final_score >= 65 else "Weak"

        suggestions = []
        if missing_skills:
            suggestions.append(f"Consider acquiring: {', '.join(missing_skills[:3])}")

        if missing_sections:
            suggestions.append(f"Add missing sections: {', '.join(missing_sections[:2])}.")

        if metric_count < 3:
            suggestions.append("Add quantified achievements (numbers, %, impact metrics).")

        if verb_count < 4:
            suggestions.append("Use stronger action verbs to improve ATS relevance.")

        suggestions.extend(formatting_suggestions[:2])

        if stuffed_keywords > 0:
            suggestions.append("Reduce repetitive keyword stuffing and use natural, role-relevant phrasing.")

        if cap_reasons:
            suggestions.extend([f"Score cap applied: {reason}" for reason in cap_reasons[:2]])

        if final_score < 70:
            suggestions.append("Align resume keywords more tightly with the target job description.")

        return AIAnalysisResult(
            ats_score=final_score,
            resume_skills=sorted(resume_skills),
            matched_skills=sorted(matched_skills),
            missing_skills=sorted(missing_skills),
            experience_match=exp_match,
            ai_suggestions=suggestions[:6],
            score_breakdown=score_breakdown,
        )

nlp_engine = NLPService()

File: resume-analyzer/backend/src/services/nlpEngine.js
import { COMMON_SKILLS } from '../utils/skillsDb.js';
import { embeddingService } from './embeddings.js';

export class NLPService {
  constructor() {
    this.skills = Array.from(COMMON_SKILLS).map((s) => s.trim().toLowerCase());
    this.requiredSections = ['summary', 'experience', 'education', 'skills', 'projects'];
    this.criticalSections = ['experience', 'skills', 'education'];
    this.actionVerbs = new Set([
      'built', 'led', 'managed', 'designed', 'developed', 'implemented', 'optimized',
      'improved', 'launched', 'scaled', 'automated', 'delivered', 'created', 'reduced',
      'increased', 'spearheaded', 'architected', 'achieved',
    ]);
    this.sectionHeadings = new Set([
      'experience', 'education', 'projects', 'summary', 'certifications',
      'achievements', 'work experience', 'profile', 'objective',
    ]);
    this.fallbackStopwords = new Set([
      'and', 'or', 'with', 'using', 'in', 'of', 'the', 'to', 'for', 'a', 'an',
      'skills', 'technologies', 'tools', 'frameworks', 'languages', 'proficient', 'knowledge',
    ]);
  }

  tokenize(text) {
    if (!text) return new Set();
    const matches = text.toLowerCase().match(/[a-zA-Z0-9+#.-]+/g) || [];
    return new Set(matches);
  }

  extractFallbackSkills(text = '') {
    const lines = text.split('\n').map((l) => l.trim()).filter(Boolean);
    const collected = [];
    let capture = false;

    for (const line of lines) {
      const lower = line.toLowerCase().replace(/^[:|\-\t]+|[:|\-\t]+$/g, '').trim();

      if (lower.includes('skill')) {
        capture = true;
        continue;
      }

      if (capture && this.sectionHeadings.has(lower)) {
        break;
      }

      if (capture) {
        collected.push(line);
      }
    }

    const sourceText = (collected.length ? collected.slice(0, 20) : lines.slice(0, 30)).join('\n');
    const candidates = [];
    const rawParts = sourceText.split(/[,|/•;\n]+/);

    for (const part of rawParts) {
      const chunk = part.replace(/\s+/g, ' ').replace(/^[-:\t]+|[-:\t]+$/g, '').trim();
      if (!chunk) continue;

      const words = chunk.match(/[A-Za-z0-9+#.-]+/g) || [];
      if (!words.length || words.length > 4) continue;

      const normalized = words.join(' ').toLowerCase();
      if (this.fallbackStopwords.has(normalized)) continue;
      if (words.some((w) => this.fallbackStopwords.has(w)) && words.length === 1) continue;
      if (!/[a-zA-Z]/.test(normalized)) continue;

      candidates.push(normalized);
    }

    const seen = new Set();
    const ordered = [];
    for (const skill of candidates) {
      if (!seen.has(skill)) {
        seen.add(skill);
        ordered.push(skill);
        if (ordered.length >= 18) break;
      }
    }

    return ordered;
  }

  extractSkills(text = '') {
    const textLower = text.toLowerCase();
    const tokens = this.tokenize(text);
    const skills = new Set();

    for (const skill of this.skills) {
      if (skill.includes(' ')) {
        if (textLower.includes(skill)) {
          skills.add(skill);
        }
      } else if (tokens.has(skill)) {
        skills.add(skill);
      }
    }

    if (skills.size < 3) {
      for (const fallback of this.extractFallbackSkills(text)) {
        skills.add(fallback);
      }
    }

    return Array.from(skills).sort();
  }

  scoreSections(resumeText = '') {
    const text = resumeText.toLowerCase();
    const found = [];
    const missing = [];

    const sectionAliases = {
      summary: ['summary', 'professional summary', 'profile', 'about', 'objective', 'about me'],
      experience: ['experience', 'work experience', 'employment', 'work history', 'professional experience', 'internship'],
      education: ['education', 'academic', 'academics', 'qualification', 'qualifications', 'degree', 'university', 'college'],
      skills: ['skills', 'technical skills', 'core competencies', 'competencies', 'technologies', 'tools', 'expertise'],
      projects: ['projects', 'personal projects', 'academic projects', 'key projects', 'portfolio', 'work samples'],
    };

    for (const [sectionKey, aliases] of Object.entries(sectionAliases)) {
      const pattern = new RegExp(`\\b(${aliases.map((a) => a.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')).join('|')})\\b`, 'i');
      if (pattern.test(text)) {
        found.push(sectionKey);
      } else {
        missing.push(sectionKey);
      }
    }

    const sectionScore = (found.length / Object.keys(sectionAliases).length) * 20;
    return { sectionScore, missingSections: missing, foundSections: found };
  }

  scoreImpactSignals(resumeText = '') {
    const text = resumeText.toLowerCase();
    const tokens = this.tokenize(text);

    const metricRegex = /\b(\d+(\.\d+)?%|\$\d+[\d,]*[kKmM]?|\d+\+?\s*(years|yrs|months|users|clients|customers|projects|features|engineers|team|members|downloads|stars|requests|tps|ms|fps|sales|revenue))\b/gi;
    const metricMatches = text.match(metricRegex) || [];
    const generalNumbers = text.match(/\b\d+([.,]\d+)?\+?\b/g) || [];
    const metricCount = metricMatches.length + Math.max(0, Math.min(Math.floor(generalNumbers.length / 3), 5));

    let verbCount = 0;
    for (const token of tokens) {
      if (this.actionVerbs.has(token)) {
        verbCount++;
      }
    }

    const metricsComponent = (Math.min(metricCount, 6) / 6) * 6;
    const verbsComponent = (Math.min(verbCount, 6) / 6) * 4;
    return { impactScore: metricsComponent + verbsComponent, metricCount, verbCount };
  }

  scoreFormatting(resumeText = '') {
    const text = resumeText || '';
    const suggestions = [];
    let score = 0.0;

    const emailPresent = /[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/.test(text);
    const phonePresent = /(\+?\d[\d\-\s]{8,}\d)/.test(text);
    const bulletMatches = text.match(/(^\s*[-*•]|\n\s*[-*•])/gm) || [];
    const bulletCount = bulletMatches.length;
    const length = text.length;

    if (emailPresent) {
      score += 3;
    } else {
      suggestions.push('Add a professional email in contact details.');
    }

    if (phonePresent) {
      score += 2;
    } else {
      suggestions.push('Include a phone number for ATS completeness.');
    }

    if (length >= 450 && length <= 9000) {
      score += 3;
    } else if (length < 450) {
      suggestions.push('Resume content is too short; add role impact and project details.');
    } else {
      suggestions.push('Resume content is too long; keep it concise and ATS-friendly.');
    }

    if (bulletCount >= 4) {
      score += 2;
    } else {
      suggestions.push('Use bullet points to improve ATS parsing and readability.');
    }

    return { formattingScore: score, formattingSuggestions: suggestions, bulletCount };
  }

  keywordStuffingPenalty(resumeText = '', jobSkills = []) {
    const text = (resumeText || '').toLowerCase();
    if (!text || !jobSkills || jobSkills.length === 0) return { penalty: 0, stuffedKeywords: 0 };

    let repeatedCount = 0;
    for (const skill of jobSkills) {
      const pattern = new RegExp(`\\b${skill.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\b`, 'gi');
      const occurrences = (text.match(pattern) || []).length;
      if (occurrences >= 6) {
        repeatedCount++;
      }
    }

    const penalty = Math.min(repeatedCount * 2.0, 8.0);
    return { penalty, stuffedKeywords: repeatedCount };
  }

  strictCaps(baseScore, jobSkillCount, skillCoverage, missingSections, metricCount, bulletCount) {
    let cappedScore = baseScore;
    const capReasons = [];

    const missingCritical = this.criticalSections.filter((s) => missingSections.includes(s));

    if (jobSkillCount > 0) {
      if (skillCoverage < 0.25) {
        cappedScore -= 15.0;
        capReasons.push('Very low role-skill coverage (<25%).');
      } else if (skillCoverage < 0.40) {
        cappedScore -= 8.0;
        capReasons.push('Low role-skill coverage (<40%).');
      }
    }

    if (missingCritical.length >= 2) {
      cappedScore -= 12.0;
      capReasons.push('Missing multiple critical sections (Experience/Skills/Education).');
    } else if (missingCritical.length === 1) {
      cappedScore -= 5.0;
      capReasons.push(`Missing ${missingCritical[0].toUpperCase()} section.`);
    }

    if (metricCount === 0) {
      cappedScore -= 6.0;
      capReasons.push('No quantifiable impact metrics detected.');
    }

    if (bulletCount < 3) {
      cappedScore -= 4.0;
      capReasons.push('Insufficient bullet structure for ATS parsing.');
    }

    return {
      cappedScore: Math.max(0.0, Math.min(100.0, cappedScore)),
      capReasons,
    };
  }

  async analyzeResumeVsJob(resumeText = '', jobDesc = '') {
    const resumeSkills = this.extractSkills(resumeText);
    const jobSkills = this.extractSkills(jobDesc);

    const resumeSkillsSet = new Set(resumeSkills);
    const jobSkillsSet = new Set(jobSkills);

    const matchedSkills = jobSkills.filter((s) => resumeSkillsSet.has(s));
    const missingSkills = jobSkills.filter((s) => !resumeSkillsSet.has(s));

    // Vector Embedding & Cosine Similarity for Semantic Relevance
    let semanticSimilarity = 0.0;
    try {
      if (resumeText.trim() && jobDesc.trim()) {
        const [vecA, vecB] = await Promise.all([
          embeddingService.generateEmbedding(resumeText),
          embeddingService.generateEmbedding(jobDesc),
        ]);
        semanticSimilarity = embeddingService.cosineSimilarity(vecA, vecB);
      }
    } catch (err) {
      console.warn('[NLP] Embedding calculation fallback:', err.message);
      semanticSimilarity = 0.5;
    }

    // ATS Rubric calculation
    let skillCoverage = 0.0;
    let skillsComponent = 0.0;
    if (jobSkills.length > 0) {
      skillCoverage = matchedSkills.length / jobSkills.length;
      skillsComponent = skillCoverage * 45;
    } else {
      skillCoverage = 0.0;
      skillsComponent = (Math.min(resumeSkills.length, 18) / 18) * 45;
    }

    const semanticComponent = semanticSimilarity * 15;
    const { sectionScore, missingSections } = this.scoreSections(resumeText);
    const { impactScore, metricCount, verbCount } = this.scoreImpactSignals(resumeText);
    const { formattingScore, formattingSuggestions, bulletCount } = this.scoreFormatting(resumeText);

    const { penalty: stuffingPenalty, stuffedKeywords } = this.keywordStuffingPenalty(resumeText, jobSkills);
    const weakImpactPenalty = metricCount < 2 ? 4.0 : 0.0;
    const weakActionPenalty = verbCount < 3 ? 2.0 : 0.0;
    const deductions = stuffingPenalty + weakImpactPenalty + weakActionPenalty;

    let finalScore = skillsComponent + semanticComponent + sectionScore + impactScore + formattingScore - deductions;
    const { cappedScore, capReasons } = this.strictCaps(
      finalScore,
      jobSkills.length,
      skillCoverage,
      missingSections,
      metricCount,
      bulletCount
    );
    finalScore = Number(Math.min(100, Math.max(0, cappedScore)).toFixed(1));

    const scoreBreakdown = {
      skills_alignment: { score: Number(skillsComponent.toFixed(1)), max: 45.0 },
      semantic_relevance: { score: Number(semanticComponent.toFixed(1)), max: 15.0 },
      section_coverage: { score: Number(sectionScore.toFixed(1)), max: 20.0 },
      impact_evidence: { score: Number(impactScore.toFixed(1)), max: 10.0 },
      formatting_quality: { score: Number(formattingScore.toFixed(1)), max: 10.0 },
      strictness_deductions: { score: Number(deductions.toFixed(1)), max: 14.0 },
      total: { score: finalScore, max: 100.0 },
    };

    const expMatch = finalScore >= 85 ? 'Strong' : finalScore >= 65 ? 'Moderate' : 'Weak';

    const suggestions = [];
    if (missingSkills.length > 0) {
      suggestions.push(`Consider acquiring: ${missingSkills.slice(0, 3).join(', ')}`);
    }
    if (missingSections.length > 0) {
      suggestions.push(`Add missing sections: ${missingSections.slice(0, 2).join(', ')}.`);
    }
    if (metricCount < 3) {
      suggestions.push('Add quantified achievements (numbers, %, impact metrics).');
    }
    if (verbCount < 4) {
      suggestions.push('Use stronger action verbs to improve ATS relevance.');
    }
    suggestions.push(...formattingSuggestions.slice(0, 2));

    if (stuffedKeywords > 0) {
      suggestions.push('Reduce repetitive keyword stuffing and use natural, role-relevant phrasing.');
    }
    if (capReasons.length > 0) {
      suggestions.push(...capReasons.slice(0, 2).map((r) => `Score cap applied: ${r}`));
    }
    if (finalScore < 70) {
      suggestions.push('Align resume keywords more tightly with the target job description.');
    }

    return {
      ats_score: finalScore,
      semantic_similarity: Number((semanticSimilarity * 100).toFixed(1)),
      resume_skills: resumeSkills,
      matched_skills: matchedSkills,
      missing_skills: missingSkills,
      experience_match: expMatch,
      ai_suggestions: suggestions.slice(0, 6),
      score_breakdown: scoreBreakdown,
    };
  }
}

export const nlpEngine = new NLPService();
